core/src/test/java/org/elasticsearch/action/admin/indices/segments/IndicesShardsStoreRequestTests.java

```diff
@@ -0,0 +1,275 @@
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAllSuccessful;
+import static org.hamcrest.Matchers.*;
+
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST)
+public class IndicesShardsStoreRequestTests extends ElasticsearchIntegrationTest {
+
+    private void indexRandomData(String index) throws ExecutionException, InterruptedException {
+        int numDocs = scaledRandomIntBetween(10, 20);
+        IndexRequestBuilder[] builders = new IndexRequestBuilder[numDocs];
+        for (int i = 0; i < builders.length; i++) {
+            builders[i] = client().prepareIndex(index, "type").setSource("field", "value");
+        }
+        indexRandom(true, builders);
+        client().admin().indices().prepareFlush().setForce(true).setWaitIfOngoing(true).execute().actionGet();
+    }
+
+    @Test
+    public void testEmpty() {
+        ensureGreen();
+        IndicesShardsStoresResponse rsp = client().admin().indices().prepareShardStores().get();
+        assertThat(rsp.getShardStatuses().size(), equalTo(0));
+    }
+
+    @Test
+    public void testBasic() throws Exception {
+        String index = "test";
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        assertAcked(prepareCreate(index).setSettings(Settings.builder()
+                        .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, "2")
+                        .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, "1")
+        ));
+        ensureGreen(index);
+        indexRandomData(index);
+        ensureGreen(index);
+
+        // no unallocated shards
+        IndicesShardsStoresResponse response = client().admin().indices().prepareShardStores(index).get();
+        assertThat(response.getShardStatuses().size(), equalTo(0));
+
+        response = client().admin().indices().shardsStores(Requests.indicesShardsStoresRequest(index).shardStatuses("all")).get();
+        assertThat(response.getShardStatuses().containsKey(index), equalTo(true));
+        ImmutableOpenIntMap<List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStores = response.getShardStatuses().get(index);
+        assertThat(shardStores.values().size(), equalTo(2));
+        for (ObjectCursor<List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStoreStatuses : shardStores.values()) {
+            for (IndicesShardsStoresResponse.ShardStoreStatus shardStoreStatus : shardStoreStatuses.value) {
+                assertThat(shardStoreStatus.getVersion(), greaterThan(-1l));
+                assertThat(shardStoreStatus.getNode(), notNullValue());
+                assertThat(shardStoreStatus.getStoreException(), nullValue());
+            }
+        }
+        ensureGreen(index);
+        logger.info("--> disable allocation");
+        disableAllocation(index);
+        logger.info("--> stop random node");
+        internalCluster().stopRandomNode(new IndexNonMasterNodePredicate(index, internalCluster().getMasterName()));
```

Comment:
`TransportIndicesShardsStoresAction` is implemented as a master-read action. I was under the impression that if the master node is stopped, then the `shardsStores` request will fail? Isn't there any chance that we hit `shardsStores` before a new master is elected?